### PR TITLE
feat(report): add support for Cosign vulnerability attestation

### PR DESCRIPTION
**Pull request by otms61**
## Description

Support `cosign-vuln` format option for [Cosing vulnerability predicate](https://github.com/sigstore/cosign/blob/main/specs/COSIGN_VULN_ATTESTATION_SPEC.md).

```
$  ./trivy image --format cosign-vuln alpine:3.10
```
<details>
<summary>Result</summary>

```json
2022-07-21T21:27:29.390+0900    INFO    Vulnerability scanning is enabled
2022-07-21T21:27:29.390+0900    INFO    Secret scanning is enabled
2022-07-21T21:27:29.390+0900    INFO    If your scanning is slow, please try '--security-checks vuln' to disable secret scanning
2022-07-21T21:27:29.390+0900    INFO    Please see also https://aquasecurity.github.io/trivy/v0.30.1-1-gf9abc93c/docs/secret/scanning/#recommendation for fast
er secret detection
2022-07-21T21:27:31.620+0900    INFO    Detected OS: alpine
2022-07-21T21:27:31.620+0900    INFO    Detecting Alpine vulnerabilities...
2022-07-21T21:27:31.621+0900    INFO    Number of language-specific files: 0
2022-07-21T21:27:31.621+0900    WARN    This OS version is no longer supported by the distribution: alpine 3.10.9
2022-07-21T21:27:31.621+0900    WARN    The vulnerability detection may be insufficient because security updates are not provided
{
  "invocation": {
    "parameters": null,
    "uri": "",
    "event_id": "",
    "builder.id": ""
  },
  "scanner": {
    "uri": "pkg:github/aquasecurity/trivy@v0.30.1-1-gf9abc93c",
    "version": "v0.30.1-1-gf9abc93c",
    "db": {
      "uri": "",
      "version": ""
    },
    "result": {
      "ArtifactName": "alpine:3.10",
      "ArtifactType": "container_image",
      "Metadata": {
        "DiffIDs": [
          "sha256:9fb3aa2f8b8023a4bebbf92aa567caf88e38e969ada9f0ac12643b2847391635"
        ],
        "ImageConfig": {
          "architecture": "amd64",
          "config": {
            "Cmd": [
              "/bin/sh"
            ],
            "Env": [
              "PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin"
            ],
            "Image": "sha256:eb2080c455e94c22ae35b3aef9e078c492a00795412e026e4d6b41ef64bc7dd8"
          },
          "container": "fdb7e80e3339e8d0599282e606c907aa5881ee4c668a68136119e6dfac6ce3a4",
          "created": "2021-04-14T19:20:05.338397761Z",
          "docker_version": "19.03.12",
          "history": [
            {
              "created": "2021-04-14T19:20:04.987219124Z",
              "created_by": "/bin/sh -c #(nop) ADD file:c5377eaa926bf412dd8d4a08b0a1f2399cfd708743533b0aa03b53d14cb4bb4e in / "
            },
            {
              "created": "2021-04-14T19:20:05.338397761Z",
              "created_by": "/bin/sh -c #(nop)  CMD [\"/bin/sh\"]",
              "empty_layer": true
            }
          ],
          "os": "linux",
          "rootfs": {
            "diff_ids": [
              "sha256:9fb3aa2f8b8023a4bebbf92aa567caf88e38e969ada9f0ac12643b2847391635"
            ],
            "type": "layers"
          }
        },
        "ImageID": "sha256:e7b300aee9f9bf3433d32bc9305bfdd22183beb59d933b48d77ab56ba53a197a",
        "OS": {
          "EOSL": true,
          "Family": "alpine",
          "Name": "3.10.9"
        },
        "RepoDigests": [
          "alpine@sha256:451eee8bedcb2f029756dc3e9d73bab0e7943c1ac55cff3a4861c52a0fdd3e98"
        ],
        "RepoTags": [
          "alpine:3.10"
        ]
      },
      "Results": [
        {
          "Class": "os-pkgs",
          "Target": "alpine:3.10 (alpine 3.10.9)",
          "Type": "alpine",
          "Vulnerabilities": [
            {
              "CVSS": {
                "nvd": {
                  "V2Score": 6.4,
                  "V2Vector": "AV:N/AC:L/Au:N/C:P/I:N/A:P",
                  "V3Score": 9.1,
                  "V3Vector": "CVSS:3.1/AV:N/AC:L/PR:N/UI:N/S:U/C:H/I:N/A:H"
                }
              },
              "CweIDs": [
                "CWE-125"
              ],
              "DataSource": {
                "ID": "alpine",
                "Name": "Alpine Secdb",
                "URL": "https://secdb.alpinelinux.org/"
              },
              "Description": "libfetch before 2021-07-26, as used in apk-tools, xbps, and other products, mishandles numeric strings for the FTP and HTTP protocols. The FTP passive mode implementation allows an out-of-bounds read because strtol is used to parse the relevant numbers into address bytes. It does not check if the line ends prematurely. If it does, the for-loop condition checks for the '\\0' terminator one byte too late.",
              "FixedVersion": "2.10.7-r0",
              "InstalledVersion": "2.10.6-r0",
              "LastModifiedDate": "2021-10-18T12:19:00Z",
              "Layer": {
                "DiffID": "sha256:9fb3aa2f8b8023a4bebbf92aa567caf88e38e969ada9f0ac12643b2847391635",
                "Digest": "sha256:396c31837116ac290458afcb928f68b6cc1c7bdd6963fc72f52f365a2a89c1b5"
              },
              "PkgName": "apk-tools",
              "PrimaryURL": "https://avd.aquasec.com/nvd/cve-2021-36159",
              "PublishedDate": "2021-08-03T14:15:00Z",
              "References": [
                "https://github.com/freebsd/freebsd-src/commits/main/lib/libfetch",
                "https://gitlab.alpinelinux.org/alpine/apk-tools/-/issues/10749",
                "https://lists.apache.org/thread.html/r61db8e7dcb56dc000a5387a88f7a473bacec5ee01b9ff3f55308aacc@%3Cdev.kafka.apache.org%3E",
                "https://lists.apache.org/thread.html/r61db8e7dcb56dc000a5387a88f7a473bacec5ee01b9ff3f55308aacc@%3Cusers.kafka.apache.org%3E",
                "https://lists.apache.org/thread.html/rbf4ce74b0d1fa9810dec50ba3ace0caeea677af7c27a97111c06ccb7@%3Cdev.kafka.apache.org%3E",
                "https://lists.apache.org/thread.html/rbf4ce74b0d1fa9810dec50ba3ace0caeea677af7c27a97111c06ccb7@%3Cusers.kafka.apache.org%3E"
              ],
              "Severity": "CRITICAL",
              "SeveritySource": "nvd",
              "VulnerabilityID": "CVE-2021-36159"
            }
          ]
        }
      ],
      "SchemaVersion": 2
    }
  },
  "metadata": {
    "scanStartedOn": "2022-07-21T21:28:33.349428+09:00",
    "scanFinishedOn": "2022-07-21T21:28:33.349428+09:00"
  }
}
```

</details>

## Related issues
- Close https://github.com/aquasecurity/trivy/issues/1646

## Checklist
- [x] I've read the [guidelines for contributing](https://aquasecurity.github.io/trivy/latest/community/contribute/pr/) to this repository.
- [x] I've followed the [conventions](https://aquasecurity.github.io/trivy/latest/community/contribute/pr/#title) in the PR title.
- [x] I've added tests that prove my fix is effective or that my feature works.
- [x] I've updated the [documentation](https://github.com/aquasecurity/trivy/blob/main/docs) with the relevant information (if needed).
- [ ] I've added usage information (if the PR introduces new options)
- [ ] I've included a "before" and "after" example to the description (if the PR is a user interface change).

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -28,7 +28,7 @@ require (
 	github.com/go-redis/redis/v8 v8.11.5
 	github.com/golang-jwt/jwt v3.2.2+incompatible
 	github.com/golang/protobuf v1.5.2
-	github.com/google/go-containerregistry v0.7.1-0.20211214010025-a65b7844a475
+	github.com/google/go-containerregistry v0.8.1-0.20220209165246-a44adc326839
```
Comment (knqyf263) [inline go.mod:31]: Do we need to bump the version?

File: go.mod [hunk 2]
```diff
@@ -49,7 +49,7 @@ require (
 	github.com/sosedoff/gitkit v0.3.0
 	github.com/spf13/cobra v1.5.0
 	github.com/spf13/pflag v1.0.5
-	github.com/spf13/viper v1.8.1
+	github.com/spf13/viper v1.12.0
```
Comment (knqyf263) [inline go.mod:52]: ditto

File: go.mod (hunk 1 above)
Comment (otms61) [reply go.mod]: No, we don't.
I updated go.mod unintentionally. I checked again, and go.mod needs no modifications.

File: go.mod (hunk 2 above)
Comment (otms61) [reply go.mod]: same
https://github.com/aquasecurity/trivy/pull/2567#discussion_r928210175

File: pkg/report/predicate/vuln.go [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+	"golang.org/x/xerrors"
+
+	"github.com/aquasecurity/trivy/pkg/clock"
+	"github.com/aquasecurity/trivy/pkg/types"
+)
+
+// CosignVulnPredicate represents the Cosign Vulnerability predicate.
+// Cosign provides the CosignVulnPredicate structure in their repository.
+// But the type of Scanner.Result is defined as map[string]interface{}, which is difficult to use,
+// so we define our own.
```
Comment (knqyf263) [inline pkg/report/predicate/vuln.go:19]: Add this link, please.
https://github.com/sigstore/cosign/pull/2096

File: pkg/report/predicate/vuln_test.go [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+	"github.com/aquasecurity/trivy/pkg/report/predicate"
+	"github.com/aquasecurity/trivy/pkg/types"
+)
+
+func TestWriter_Write(t *testing.T) {
+	testCases := []struct {
+		name          string
+		detectedVulns []types.DetectedVulnerability
+		want          predicate.CosignVulnPredicate
+		wantResult    types.Report
```
Comment (knqyf263) [inline pkg/report/predicate/vuln_test.go:24]: Do we need it?

File: pkg/report/predicate/vuln_test.go [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+	"github.com/stretchr/testify/assert"
+
+	"github.com/aquasecurity/trivy/pkg/clock"
+	ftypes "github.com/aquasecurity/trivy/pkg/fanal/types"
+	"github.com/aquasecurity/trivy/pkg/report/predicate"
+	"github.com/aquasecurity/trivy/pkg/types"
+)
+
+func TestWriter_Write(t *testing.T) {
+	testCases := []struct {
```
Comment (knqyf263) [inline pkg/report/predicate/vuln_test.go:20]: It is a nitpick, but we're recently trying to comply with the same convention.
https://github.com/aquasecurity/trivy/blob/d93a9978005889638472ff198c4af6ef56b1246c/pkg/licensing/classifier_test.go#L16

```suggestion
	tests := []struct {
```

File: pkg/report/predicate/vuln_test.go [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+				},
+				Metadata: predicate.Metadata{
+					ScanStartedOn:  time.Date(2022, time.July, 22, 12, 20, 30, 5, time.UTC),
+					ScanFinishedOn: time.Date(2022, time.July, 22, 12, 20, 30, 5, time.UTC),
+				},
+			},
+		},
+	}
+
+	for _, tc := range testCases {
```
Comment (knqyf263) [inline pkg/report/predicate/vuln_test.go:83]: nit: ditto

```suggestion
	for _, tt := range testCases {
```

File: pkg/report/predicate/vuln_test.go [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+				},
+			}
+
+			output := bytes.NewBuffer(nil)
+
+			clock.SetFakeTime(t, time.Date(2022, 7, 22, 12, 20, 30, 5, time.UTC))
+			writer := predicate.NewWriter(output, "dev")
+
+			err := writer.Write(inputResults)
+			assert.NoError(t, err)
```
Comment (knqyf263) [inline pkg/report/predicate/vuln_test.go:103]: nit: IMHO, `require` is better here as all the subsequent assertions will fail anyway when it returns an error.

```suggestion
			require.NoError(t, err)
```

File: pkg/report/predicate/vuln_test.go [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+
+			var got predicate.CosignVulnPredicate
+			err = json.Unmarshal(output.Bytes(), &got)
+			assert.NoError(t, err, "invalid json written")
```
Comment (knqyf263) [inline pkg/report/predicate/vuln_test.go:107]: ditto

File: docs/docs/attestation/vuln.md [hunk 9]
```diff
@@ -0,0 +1,167 @@
...
+## Create Cosign Vulnerability Attestation
+
+[Cosign](https://github.com/sigstore/cosign) supports generating and verifying [in-toto attestations](https://github.com/in-toto/attestation). This tool enables you to sign and verify Cosign vulnerability attestation.
+
+!!! note
+    In the following examples, the `cosign` command will write an attestation to a target OCI registry, so you must have permission to write.
+    If you want to avoid writing an OCI registry and only want to see an attestation, add the `--no-upload` option to the `cosign` command.
+
+
+Cosign can generate key pairs and use them for signing and verification. Read more about [how to generate key pairs](https://docs.sigstore.dev/cosign/key-generation).
```
Comment (knqyf263) [inline docs/docs/attestation/vuln.md:154]: Can we add a keyless section?

File: docs/docs/attestation/vuln.md [hunk 10]
```diff
@@ -0,0 +1,167 @@
+# Cosign Vulnerability Attestation
+
+## Generate Cosign Vulnerability Predicate 
+
+Trivy generates reports in the [Cosign vulnerability predicate format](https://github.com/sigstore/cosign/blob/95b74db89941e8ec85e768f639efd4d948db06cd/specs/COSIGN_VULN_ATTESTATION_SPEC.md).
```
Comment (knqyf263) [inline docs/docs/attestation/vuln.md:5]:
```suggestion
Trivy generates reports in the [Cosign vulnerability predicate format][vuln-attest-spec]
```

And put it at the bottom.

```
[vuln-attest-spec]: https://github.com/sigstore/cosign/blob/95b74db89941e8ec85e768f639efd4d948db06cd/specs/COSIGN_VULN_ATTESTATION_SPEC.md
```

File: pkg/report/predicate/vuln.go (hunk 3 above)
Comment (otms61) [reply pkg/report/predicate/vuln.go:19]: Alright. I've added the link.

File: pkg/report/predicate/vuln_test.go (hunk 4 above)
Comment (otms61) [reply pkg/report/predicate/vuln_test.go]: Sorry, I forgot to remove it.

File: pkg/report/predicate/vuln_test.go (hunk 5 above)
Comment (otms61) [reply pkg/report/predicate/vuln_test.go]: I've fixed it.

File: pkg/report/predicate/vuln_test.go (hunk 6 above)
Comment (otms61) [reply pkg/report/predicate/vuln_test.go]: I've fixed it.

File: pkg/report/predicate/vuln_test.go (hunk 7 above)
Comment (otms61) [reply pkg/report/predicate/vuln_test.go]: It makes sense. I've fixed it.

File: pkg/report/predicate/vuln_test.go (hunk 8 above)
Comment (otms61) [reply pkg/report/predicate/vuln_test.go]: I've fixed it.

File: docs/docs/attestation/vuln.md (hunk 9 above)
Comment (otms61) [reply docs/docs/attestation/vuln.md:156]: Alright. I've added a keyless signing section.

File: docs/docs/attestation/vuln.md (hunk 10 above)
Comment (otms61) [reply docs/docs/attestation/vuln.md]: I've fixed it.

File: mkdocs.yml [hunk 11]
```diff
@@ -74,6 +74,7 @@ nav:
           - SPDX: docs/sbom/spdx.md
       - Attestation:
           - SBOM: docs/attestation/sbom.md
+          - Cosign Vulnerability Predicate: docs/attestation/vuln.md
```
Comment (knqyf263) [inline mkdocs.yml:77]: To align with their doc.

>Cosign Vulnerability Scan Record Attestation Specification

https://github.com/sigstore/cosign/blob/main/specs/COSIGN_VULN_ATTESTATION_SPEC.md

```suggestion
          - Cosign Vulnerability Scan Record: docs/attestation/vuln.md
```

File: pkg/report/predicate/vuln.go [hunk 12]
```diff
@@ -0,0 +1,91 @@
...
+	DB      DB           `json:"db"`
+	Result  types.Report `json:"result"`
+}
+
+type Metadata struct {
+	ScanStartedOn  time.Time `json:"scanStartedOn"`
+	ScanFinishedOn time.Time `json:"scanFinishedOn"`
+}
+
+type Writer struct {
```
Comment (knqyf263) [inline pkg/report/predicate/vuln.go:52]: We may have more predicates in the future.

```suggestion
type VulnWriter struct {
```

File: pkg/report/predicate/vuln.go [hunk 13]
```diff
@@ -0,0 +1,91 @@
...
+	output  io.Writer
+	version string
+}
+
+func NewWriter(output io.Writer, version string) Writer {
```
Comment (knqyf263) [inline pkg/report/predicate/vuln.go:57]:
```suggestion
func NewVulnWriter(output io.Writer, version string) Writer {
```

File: pkg/report/predicate/vuln.go [hunk 14]
```diff
@@ -0,0 +1,91 @@
...
+	"github.com/aquasecurity/trivy/pkg/clock"
+	"github.com/aquasecurity/trivy/pkg/types"
+)
+
+// CosignVulnPredicate represents the Cosign Vulnerability predicate.
+// Cosign provides the CosignVulnPredicate structure in their repository.
+// But the type of Scanner.Result is defined as map[string]interface{}, which is difficult to use,
+// so we define our own.
+// The PR is in progress to replace Scanner.Result type to interface{}.
+// https://github.com/sigstore/cosign/pull/2096
```
Comment (knqyf263) [inline pkg/report/predicate/vuln.go:21]: This PR got merged.

File: pkg/report/predicate/vuln.go (hunk 12 above)
Comment (otms61) [reply pkg/report/predicate/vuln.go]: Done

File: pkg/report/predicate/vuln.go (hunk 13 above)
Comment (otms61) [reply pkg/report/predicate/vuln.go]: Done

File: pkg/report/predicate/vuln.go (hunk 14 above)
Comment (otms61) [reply pkg/report/predicate/vuln.go]: I have added comments on the background and references.

**Review by knqyf263 — APPROVED** [review]

**Pull request merged**